File: src/minimax_studio/workflows/stitch.py
from __future__ import annotations

import shutil
from collections.abc import Callable
from pathlib import Path

from minimax_studio.files import ensure_dir
from minimax_studio.media import compose_final_video, concat_videos, get_duration_seconds, pad_video_to_duration
from minimax_studio.schemas import StitchOptions, StitchResult
from minimax_studio.workflows import emit_report


def run_stitch(
    options: StitchOptions,
    *,
    reporter: Callable[[str], None] | None = None,
) -> StitchResult:
    validate_stitch_options(options)

    output_dir = ensure_dir(options.output_path.resolve().parent)

    stitched_video_path = output_dir / "edit.mp4"
    padded_video_path = output_dir / "timed.mp4"
    final_video_path = options.output_path.resolve()

    emit_report(reporter, "step 1/3: stitching source video...")
    concat_videos(options.video_paths, stitched_video_path)

    narration_duration = get_duration_seconds(options.narration_path)
    video_duration = get_duration_seconds(stitched_video_path)
    target_duration = max(video_duration, narration_duration)

    emit_report(reporter, "step 2/3: padding video to target duration...")
    pad_video_to_duration(stitched_video_path, padded_video_path, target_duration)
    emit_report(
        reporter,
        f"stitched video duration={video_duration:.2f}s, narration duration={narration_duration:.2f}s, target={target_duration:.2f}s",
    )

    emit_report(reporter, "step 3/3: composing final video...")
    compose_final_video(
        video_path=padded_video_path,
        narration_path=options.narration_path,
        music_path=options.music_path,
        output_path=final_video_path,
        target_duration=target_duration,
    )
    emit_report(reporter, f"final video saved to: {final_video_path}")

    return StitchResult(
        stitched_video_path=stitched_video_path,
        padded_video_path=padded_video_path,
        final_video_path=final_video_path,
    )


def validate_stitch_options(options: StitchOptions) -> None:
    if shutil.which("ffmpeg") is None or shutil.which("ffprobe") is None:
        raise SystemExit("ffmpeg and ffprobe are required")

    if not options.video_paths:
        raise SystemExit("at least one input video is required")

    missing_video = [path for path in options.video_paths if not path.exists()]
    if missing_video:
        raise SystemExit(f"input video does not exist: {missing_video[0]}")

    if not options.narration_path.exists():
        raise SystemExit(f"narration file does not exist: {options.narration_path}")

    if options.music_path is not None and not options.music_path.exists():
        raise SystemExit(f"music file does not exist: {options.music_path}")

File: internal/workflows/stitch.go
package workflows

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"

	"github.com/minimax-ai/minimax-studio/internal/media"
	"github.com/minimax-ai/minimax-studio/internal/schemas"
)

// StitchWorkflow 素材合成工作流
type StitchWorkflow struct {
}

// NewStitchWorkflow 创建 StitchWorkflow
func NewStitchWorkflow() *StitchWorkflow {
	return &StitchWorkflow{}
}

// Run 执行素材合成
func (w *StitchWorkflow) Run(opts schemas.StitchOptions, reporter Reporter) (*schemas.StitchResult, error) {
	if reporter == nil {
		reporter = func(stage string) {}
	}

	// 验证 ffmpeg
	if !isFFmpegAvailable() {
		return nil, fmt.Errorf("ffmpeg and ffprobe are required")
	}

	// 验证输入
	if len(opts.VideoPaths) == 0 {
		return nil, fmt.Errorf("at least one input video is required")
	}

	for _, p := range opts.VideoPaths {
		if _, err := os.Stat(p); os.IsNotExist(err) {
			return nil, fmt.Errorf("input video does not exist: %s", p)
		}
	}

	if _, err := os.Stat(opts.NarrationPath); os.IsNotExist(err) {
		return nil, fmt.Errorf("narration file does not exist: %s", opts.NarrationPath)
	}

	if opts.MusicPath != "" {
		if _, err := os.Stat(opts.MusicPath); os.IsNotExist(err) {
			return nil, fmt.Errorf("music file does not exist: %s", opts.MusicPath)
		}
	}

	// 确保输出目录存在
	outputDir := filepath.Dir(opts.OutputPath)
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create output dir: %w", err)
	}

	stitchedVideoPath := filepath.Join(outputDir, "edit.mp4")
	paddedVideoPath := filepath.Join(outputDir, "timed.mp4")
	finalVideoPath := opts.OutputPath

	// Step 1: 拼接视频
	reporter("step 1/3: stitching source video...")
	if err := media.ConcatVideos(opts.VideoPaths, stitchedVideoPath); err != nil {
		return nil, fmt.Errorf("failed to concat videos: %w", err)
	}

	// Step 2: 计算时长并填充
	reporter("step 2/3: padding video to target duration...")
	narrationDuration, err := media.GetDurationSeconds(opts.NarrationPath)
	if err != nil {
		return nil, fmt.Errorf("failed to get narration duration: %w", err)
	}

	videoDuration, err := media.GetDurationSeconds(stitchedVideoPath)
	if err != nil {
		return nil, fmt.Errorf("failed to get video duration: %w", err)
	}

	targetDuration := narrationDuration
	if videoDuration > narrationDuration {
		targetDuration = videoDuration
	}

	if err := media.PadVideoToDuration(stitchedVideoPath, paddedVideoPath, targetDuration); err != nil {
		return nil, fmt.Errorf("failed to pad video: %w", err)
	}

	reporter(fmt.Sprintf(
		"stitched video duration=%.2fs, narration duration=%.2fs, target=%.2fs",
		videoDuration, narrationDuration, targetDuration,
	))

	// Step 3: 合成最终视频
	reporter("step 3/3: composing final video...")
	if err := media.ComposeFinalVideo(paddedVideoPath, opts.NarrationPath, opts.MusicPath, finalVideoPath, targetDuration); err != nil {
		return nil, fmt.Errorf("failed to compose final video: %w", err)
	}
	reporter(fmt.Sprintf("final video saved to: %s", finalVideoPath))

	return &schemas.StitchResult{
		StitchedVideoPath: stitchedVideoPath,
		PaddedVideoPath:   paddedVideoPath,
		FinalVideoPath:    finalVideoPath,
	}, nil
}

func isFFmpegAvailable() bool {
	cmd := exec.Command("ffmpeg", "-version")
	return cmd.Run() == nil
}
